File: lib/converter.js
import { randomUUID } from 'crypto';

// ─────────────────────────────────────────────
// 检测客户端格式
// ─────────────────────────────────────────────
export function detectClientFormat(url) {
  const path = url.split('?')[0];
  if (path.endsWith('/v1/messages'))           return 'anthropic';
  if (path.endsWith('/v1/chat/completions'))   return 'openai-chat';
  if (path.endsWith('/v1/responses'))          return 'openai-responses';
  return null; // 不拦截
}

// ─────────────────────────────────────────────
// → 内部格式（统一用 Anthropic-like 结构）
// ─────────────────────────────────────────────

/**
 * Anthropic 请求 → 内部格式（几乎原样保留）
 */
export function fromAnthropic(body) {
  return {
    model:     body.model,
    messages:  body.messages  || [],
    system:    body.system    || null,
    stream:    body.stream    || false,
    maxTokens: body.max_tokens || 4096,
    tools:     body.tools     || null,
    extra:     body,   // 保留原始字段（temperature 等）
  };
}

/**
 * OpenAI Chat 请求 → 内部格式
 */
export function fromOpenAIChat(body) {
  const messages = [];
  let system = null;

  for (const msg of (body.messages || [])) {
    if (msg.role === 'system') {
      system = typeof msg.content === 'string' ? msg.content : msg.content?.[0]?.text || '';
      continue;
    }
    messages.push({
      role:    msg.role,
      content: normalizeOpenAIContent(msg.content),
    });
  }

  return {
    model:     body.model,
    messages,
    system,
    stream:    body.stream     || false,
    maxTokens: body.max_tokens || 4096,
    tools:     body.tools      || null,
    extra:     body,
  };
}

/**
 * OpenAI Responses API 请求 → 内部格式
 */
export function fromOpenAIResponses(body) {
  const messages = [];

  // input 可以是字符串或数组
  const input = Array.isArray(body.input)
    ? body.input
    : (typeof body.input === 'string' ? [{ role: 'user', content: body.input }] : []);

  for (const item of input) {
    if (item.type === 'message' || item.role) {
      messages.push({
        role:    item.role,
        content: normalizeResponsesContent(item.content),
      });
    }
  }

  return {
    model:     body.model,
    messages,
    system:    body.instructions || null,
    stream:    body.stream        || false,
    maxTokens: body.max_output_tokens || 4096,
    tools:     body.tools         || null,
    extra:     body,
  };
}

// ─────────────────────────────────────────────
// 内部格式 → 上游 Provider 格式
// ─────────────────────────────────────────────

/**
 * 内部格式 → Anthropic 请求体
 */
export function toAnthropicBody(internal, targetModel) {
  const body = {
    model:      targetModel,
    messages:   internal.messages,
    max_tokens: internal.maxTokens,
    stream:     internal.stream,
  };
  if (internal.system) body.system = internal.system;
  if (internal.tools)  body.tools  = internal.tools;
  // 透传 temperature 等额外字段
  for (const k of ['temperature', 'top_p', 'top_k', 'stop_sequences', 'metadata']) {
    if (internal.extra[k] !== undefined) body[k] = internal.extra[k];
  }
  return body;
}

/**
 * 内部格式 → OpenAI Chat 请求体
 */
export function toOpenAIChatBody(internal, targetModel) {
  const messages = [];
  if (internal.system) messages.push({ role: 'system', content: internal.system });

  for (const msg of internal.messages) {
    messages.push({
      role:    msg.role,
      content: toOpenAIContent(msg.content),
    });
  }

  const body = {
    model:      targetModel,
    messages,
    max_tokens: internal.maxTokens,
    stream:     internal.stream,
  };
  if (internal.tools) body.tools = internal.tools;
  for (const k of ['temperature', 'top_p', 'frequency_penalty', 'presence_penalty']) {
    if (internal.extra[k] !== undefined) body[k] = internal.extra[k];
  }
  return body;
}

// ─────────────────────────────────────────────
// 响应转换（非流式）
// ─────────────────────────────────────────────

/**
 * Anthropic 响应 → 指定客户端格式
 */
export function convertAnthropicResponse(data, clientFormat, originalModel) {
  if (clientFormat === 'anthropic') return data;
  if (clientFormat === 'openai-chat') return anthropicToOpenAIChatResponse(data, originalModel);
  if (clientFormat === 'openai-responses') return anthropicToResponsesResponse(data, originalModel);
  return data;
}

/**
 * OpenAI Chat 响应 → 指定客户端格式
 */
export function convertOpenAIChatResponse(data, clientFormat, originalModel) {
  if (clientFormat === 'openai-chat') return data;
  if (clientFormat === 'anthropic') return openAIChatToAnthropicResponse(data, originalModel);
  if (clientFormat === 'openai-responses') return openAIChatToResponsesResponse(data, originalModel);
  return data;
}

// ─────────────────────────────────────────────
// SSE 流式转换
// ─────────────────────────────────────────────

/**
 * 创建流式 SSE 转换器
 * 返回 transform(line) 函数，把上游 SSE 行转成目标格式的行（数组）
 */
export function createStreamTransformer(providerFormat, clientFormat) {
  // 相同格式直接透传
  if (providerFormat === clientFormat ||
      (providerFormat === 'openai' && clientFormat === 'openai-chat')) {
    return (line) => [line];
  }

  const state = { msgId: `chatcmpl-${randomUUID().slice(0,8)}`, respId: `resp_${randomUUID().slice(0,8)}`, model: '', started: false, done: false };

  if (providerFormat === 'anthropic' && clientFormat === 'openai-chat') {
    return makeAnthropicToChat(state);
  }
  if (providerFormat === 'anthropic' && clientFormat === 'openai-responses') {
    return makeAnthropicToResponses(state);
  }
  if (providerFormat === 'openai' && clientFormat === 'anthropic') {
    return makeOpenAIToAnthropic(state);
  }
  if (providerFormat === 'openai' && clientFormat === 'openai-responses') {
    return makeOpenAIToResponses(state);
  }

  return (line) => [line]; // fallback
}

// ─────────────────────────────────────────────
// 内部工具函数
// ─────────────────────────────────────────────

function normalizeOpenAIContent(content) {
  if (typeof content === 'string') return [{ type: 'text', text: content }];
  if (!Array.isArray(content)) return [];
  return content.map(block => {
    if (block.type === 'text')       return { type: 'text', text: block.text };
    if (block.type === 'image_url')  return openAIImageToAnthropic(block.image_url);
    return block;
  });
}

function normalizeResponsesContent(content) {
  if (typeof content === 'string') return [{ type: 'text', text: content }];
  if (!Array.isArray(content)) return [];
  return content.map(block => {
    if (block.type === 'input_text')  return { type: 'text', text: block.text };
    if (block.type === 'input_image') return responsesImageToAnthropic(block);
    if (block.type === 'output_text') return { type: 'text', text: block.text };
    return block;
  });
}

function openAIImageToAnthropic(imageUrl) {
  const url = typeof imageUrl === 'string' ? imageUrl : imageUrl.url;
  const m = url.match(/^data:([^;]+);base64,(.+)$/);
  if (m) return { type: 'image', source: { type: 'base64', media_type: m[1], data: m[2] } };
  return { type: 'image', source: { type: 'url', url } };
}

function responsesImageToAnthropic(block) {
  if (block.image_url) return openAIImageToAnthropic(block.image_url);
  if (block.url)       return { type: 'image', source: { type: 'url', url: block.url } };
  return block;
}

function toOpenAIContent(content) {
  if (!Array.isArray(content)) return content;
  if (content.every(b => b.type === 'text')) {
    return content.map(b => b.text).join('\n');
  }
  return content.map(block => {
    if (block.type === 'text')  return { type: 'text', text: block.text };
    if (block.type === 'image') {
      const src = block.source;
      const url = src.type === 'base64' ? `data:${src.media_type};base64,${src.data}` : src.url;
      return { type: 'image_url', image_url: { url } };
    }
    return block;
  });
}

function extractText(content) {
  if (!Array.isArray(content)) return String(content || '');
  return content.filter(b => b.type === 'text').map(b => b.text).join('');
}

// ── 非流式响应转换 ──────────────────────────

function anthropicToOpenAIChatResponse(data, model) {
  return {
    id:      `chatcmpl-${randomUUID().slice(0,8)}`,
    object:  'chat.completion',
    created: Math.floor(Date.now() / 1000),
    model,
    choices: [{
      index:         0,
      message:       { role: 'assistant', content: extractText(data.content) },
      finish_reason: data.stop_reason === 'end_turn' ? 'stop' : (data.stop_reason || 'stop'),
    }],
    usage: {
      prompt_tokens:     data.usage?.input_tokens  || 0,
      completion_tokens: data.usage?.output_tokens || 0,
      total_tokens:      (data.usage?.input_tokens || 0) + (data.usage?.output_tokens || 0),
    },
  };
}

function anthropicToResponsesResponse(data, model) {
  const text = extractText(data.content);
  return {
    id:         `resp_${randomUUID().slice(0,8)}`,
    object:     'response',
    created_at: Math.floor(Date.now() / 1000),
    model,
    status:     'completed',
    output: [{
      type:    'message',
      id:      `msg_${randomUUID().slice(0,8)}`,
      role:    'assistant',
      content: [{ type: 'output_text', text }],
    }],
    usage: {
      input_tokens:  data.usage?.input_tokens  || 0,
      output_tokens: data.usage?.output_tokens || 0,
    },
  };
}

function openAIChatToAnthropicResponse(data, model) {
  const choice = data.choices?.[0] || {};
  const text   = choice.message?.content || '';
  return {
    id:           `msg_${randomUUID().slice(0,8)}`,
    type:         'message',
    role:         'assistant',
    model,
    content:      [{ type: 'text', text }],
    stop_reason:  choice.finish_reason === 'stop' ? 'end_turn' : choice.finish_reason,
    usage: {
      input_tokens:  data.usage?.prompt_tokens     || 0,
      output_tokens: data.usage?.completion_tokens || 0,
    },
  };
}

function openAIChatToResponsesResponse(data, model) {
  const choice = data.choices?.[0] || {};
  const text   = choice.message?.content || '';
  return {
    id:         `resp_${randomUUID().slice(0,8)}`,
    object:     'response',
    created_at: Math.floor(Date.now() / 1000),
    model,
    status:     'completed',
    output: [{
      type: 'message', id: `msg_${randomUUID().slice(0,8)}`, role: 'assistant',
      content: [{ type: 'output_text', text }],
    }],
    usage: {
      input_tokens:  data.usage?.prompt_tokens     || 0,
      output_tokens: data.usage?.completion_tokens || 0,
    },
  };
}

// ── 流式 SSE 转换器工厂 ──────────────────────

function parseSseLine(line) {
  if (line.startsWith('data: ')) {
    const raw = line.slice(6);
    if (raw === '[DONE]') return { done: true };
    try { return { data: JSON.parse(raw) }; } catch { return null; }
  }
  return null;
}

function sse(event, data) {
  return `event: ${event}\ndata: ${JSON.stringify(data)}\n\n`;
}

function chatDelta(id, model, delta, finishReason = null) {
  return `data: ${JSON.stringify({
    id, object: 'chat.completion.chunk',
    created: Math.floor(Date.now() / 1000), model,
    choices: [{ index: 0, delta, finish_reason: finishReason }],
  })}\n\n`;
}

function makeAnthropicToChat(s) {
  return (line) => {
    const p = parseSseLine(line);
    if (!p) return [];
    if (p.done) return ['data: [DONE]\n\n'];
    const { data } = p;
    if (!data) return [];

    if (data.type === 'message_start') {
      s.model = data.message?.model || '';
      return [chatDelta(s.msgId, s.model, { role: 'assistant' })];
    }
    if (data.type === 'content_block_delta' && data.delta?.type === 'text_delta') {
      return [chatDelta(s.msgId, s.model, { content: data.delta.text })];
    }
    if (data.type === 'message_delta' && data.delta?.stop_reason) {
      const fr = data.delta.stop_reason === 'end_turn' ? 'stop' : data.delta.stop_reason;
      return [chatDelta(s.msgId, s.model, {}, fr), 'data: [DONE]\n\n'];
    }
    return [];
  };
}

function makeAnthropicToResponses(s) {
  const itemId = `msg_${randomUUID().slice(0,8)}`;
  return (line) => {
    const p = parseSseLine(line);
    if (!p) return [];
    if (p.done) return [];
    const { data } = p;
    if (!data) return [];

    if (data.type === 'message_start') {
      s.model = data.message?.model || '';
      return [
        sse('response.created', { type: 'response.created', response: { id: s.respId, object: 'response', status: 'in_progress', model: s.model, output: [] } }),
        sse('response.output_item.added', { type: 'response.output_item.added', output_index: 0, item: { type: 'message', id: itemId, role: 'assistant', content: [] } }),
        sse('response.content_part.added', { type: 'response.content_part.added', item_id: itemId, output_index: 0, content_index: 0, part: { type: 'output_text', text: '' } }),
      ];
    }
    if (data.type === 'content_block_delta' && data.delta?.type === 'text_delta') {
      return [sse('response.output_text.delta', { type: 'response.output_text.delta', item_id: itemId, output_index: 0, content_index: 0, delta: data.delta.text })];
    }
    if (data.type === 'message_stop') {
      return [
        sse('response.output_text.done',  { type: 'response.output_text.done', item_id: itemId, output_index: 0, content_index: 0, text: '' }),
        sse('response.output_item.done',  { type: 'response.output_item.done', output_index: 0, item: { type: 'message', id: itemId, role: 'assistant', content: [{ type: 'output_text', text: '' }] } }),
        sse('response.completed',         { type: 'response.completed', response: { id: s.respId, object: 'response', status: 'completed', model: s.model } }),
      ];
    }
    return [];
  };
}

function makeOpenAIToAnthropic(s) {
  let blockStarted = false;
  return (line) => {
    const p = parseSseLine(line);
    if (!p) return [];
    if (p.done) {
      return [
        `event: message_stop\ndata: {"type":"message_stop"}\n\n`,
      ];
    }
    const { data } = p;
    if (!data) return [];
    const delta   = data.choices?.[0]?.delta || {};
    const finish  = data.choices?.[0]?.finish_reason;
    const out     = [];

    if (!s.started) {
      s.model = data.model || '';
      out.push(`event: message_start\ndata: ${JSON.stringify({ type: 'message_start', message: { id: `msg_${randomUUID().slice(0,8)}`, type: 'message', role: 'assistant', model: s.model, content: [], stop_reason: null, usage: { input_tokens: 0, output_tokens: 0 } } })}\n\n`);
      s.started = true;
    }
    if (delta.content !== undefined && !blockStarted) {
      out.push(`event: content_block_start\ndata: {"type":"content_block_start","index":0,"content_block":{"type":"text","text":""}}\n\n`);
      blockStarted = true;
    }
    if (delta.content) {
      out.push(`event: content_block_delta\ndata: ${JSON.stringify({ type: 'content_block_delta', index: 0, delta: { type: 'text_delta', text: delta.content } })}\n\n`);
    }
    if (finish) {
      out.push(`event: content_block_stop\ndata: {"type":"content_block_stop","index":0}\n\n`);
      out.push(`event: message_delta\ndata: ${JSON.stringify({ type: 'message_delta', delta: { stop_reason: finish === 'stop' ? 'end_turn' : finish, stop_sequence: null }, usage: { output_tokens: 0 } })}\n\n`);
    }
    return out;
  };
}

function makeOpenAIToResponses(s) {
  const itemId = `msg_${randomUUID().slice(0,8)}`;
  return (line) => {
    const p = parseSseLine(line);
    if (!p) return [];
    if (p.done) {
      return [
        sse('response.output_item.done', { type: 'response.output_item.done', output_index: 0, item: { type: 'message', id: itemId, role: 'assistant', content: [] } }),
        sse('response.completed', { type: 'response.completed', response: { id: s.respId, status: 'completed', model: s.model } }),
      ];
    }
    const { data } = p;
    if (!data) return [];
    const delta  = data.choices?.[0]?.delta || {};
    const finish = data.choices?.[0]?.finish_reason;
    const out    = [];

    if (!s.started) {
      s.model = data.model || '';
      out.push(sse('response.created', { type: 'response.created', response: { id: s.respId, object: 'response', status: 'in_progress', model: s.model, output: [] } }));
      out.push(sse('response.output_item.added', { type: 'response.output_item.added', output_index: 0, item: { type: 'message', id: itemId, role: 'assistant', content: [] } }));
      out.push(sse('response.content_part.added', { type: 'response.content_part.added', item_id: itemId, output_index: 0, content_index: 0, part: { type: 'output_text', text: '' } }));
      s.started = true;
    }
    if (delta.content) {
      out.push(sse('response.output_text.delta', { type: 'response.output_text.delta', item_id: itemId, output_index: 0, content_index: 0, delta: delta.content }));
    }
    if (finish) {
      out.push(sse('response.output_text.done', { type: 'response.output_text.done', item_id: itemId, output_index: 0, content_index: 0, text: '' }));
    }
    return out;
  };
}

File: go/internal/converter/converter.go
// Package converter 对齐 Node 版 lib/converter.js：三种 API 格式（Anthropic / OpenAI Chat /
// OpenAI Responses）的请求/响应/流式双向转换。
//
// 内部统一用 Anthropic-like 结构表示；为贴合 Node 的动态语义，JSON 体一律用 map[string]any / []any，
// 而非强类型结构体，以最大程度保证与原实现行为等价（这是迁移的最大风险点，需对拍测试）。
package converter

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"strings"
)

// Internal 内部统一格式（Anthropic-like）
type Internal struct {
	Model     string
	Messages  []any          // 每个元素为 map[string]any: {role, content}
	System    any            // string 或 nil
	Stream    bool
	MaxTokens int
	Tools     any            // 数组或 nil
	Extra     map[string]any // 原始请求体，用于透传 temperature 等
}

// DetectClientFormat 按端点路径识别客户端格式（对齐 Node 版）。
func DetectClientFormat(path string) string {
	p := strings.Split(path, "?")[0]
	switch {
	case strings.HasSuffix(p, "/v1/messages"):
		return "anthropic"
	case strings.HasSuffix(p, "/v1/chat/completions"):
		return "openai-chat"
	case strings.HasSuffix(p, "/v1/responses"):
		return "openai-responses"
	}
	return ""
}

// IsPassthrough 判断 provider 和 client 格式是否可直接透传（不需逐行转换）。
// 供 proxy 和 stream 模块共用，消除判定逻辑分歧。
func IsPassthrough(providerFormat, clientFormat string) bool {
	return providerFormat == clientFormat ||
		(providerFormat == "openai" && clientFormat == "openai-chat")
}

// ── 客户端请求 → 内部格式 ──────────────────────

// FromAnthropic Anthropic 请求 → 内部格式（几乎原样保留）。
func FromAnthropic(body map[string]any) *Internal {
	return &Internal{
		Model:     getString(body, "model"),
		Messages:  getSlice(body, "messages"),
		System:    body["system"],
		Stream:    getBool(body, "stream"),
		MaxTokens: getIntDefault(body, "max_tokens", 4096),
		Tools:     body["tools"],
		Extra:     body,
	}
}

// FromOpenAIChat OpenAI Chat 请求 → 内部格式。
func FromOpenAIChat(body map[string]any) *Internal {
	var messages []any
	var system any

	for _, m := range getSlice(body, "messages") {
		msg, ok := m.(map[string]any)
		if !ok {
			continue
		}
		role := getString(msg, "role")
		if role == "system" {
			system = systemTextFromContent(msg["content"])
			continue
		}
		messages = append(messages, map[string]any{
			"role":    role,
			"content": normalizeOpenAIContent(msg["content"]),
		})
	}

	return &Internal{
		Model:     getString(body, "model"),
		Messages:  messages,
		System:    system,
		Stream:    getBool(body, "stream"),
		MaxTokens: getIntDefault(body, "max_tokens", 4096),
		Tools:     body["tools"],
		Extra:     body,
	}
}

// FromOpenAIResponses OpenAI Responses 请求 → 内部格式。
func FromOpenAIResponses(body map[string]any) *Internal {
	var input []any
	switch v := body["input"].(type) {
	case []any:
		input = v
	case string:
		input = []any{map[string]any{"role": "user", "content": v}}
	}

	var messages []any
	for _, it := range input {
		item, ok := it.(map[string]any)
		if !ok {
			continue
		}
		if item["type"] == "message" || item["role"] != nil {
			messages = append(messages, map[string]any{
				"role":    item["role"],
				"content": normalizeResponsesContent(item["content"]),
			})
		}
	}

	var system any
	if v, ok := body["instructions"]; ok {
		system = v
	}

	return &Internal{
		Model:     getString(body, "model"),
		Messages:  messages,
		System:    system,
		Stream:    getBool(body, "stream"),
		MaxTokens: getIntDefault(body, "max_output_tokens", 4096),
		Tools:     body["tools"],
		Extra:     body,
	}
}

// ── 内部格式 → 上游 Provider 请求体 ───────────

// ToAnthropicBody 内部格式 → Anthropic 请求体。
func ToAnthropicBody(in *Internal, targetModel string) map[string]any {
	body := map[string]any{
		"model":      targetModel,
		"messages":   in.Messages,
		"max_tokens": in.MaxTokens,
		"stream":     in.Stream,
	}
	if in.System != nil && in.System != "" {
		body["system"] = in.System
	}
	if in.Tools != nil {
		body["tools"] = in.Tools
	}
	for _, k := range []string{"temperature", "top_p", "top_k", "stop_sequences", "metadata"} {
		if v, ok := in.Extra[k]; ok {
			body[k] = v
		}
	}
	return body
}

// ToOpenAIChatBody 内部格式 → OpenAI Chat 请求体。
func ToOpenAIChatBody(in *Internal, targetModel string) map[string]any {
	var messages []any
	if s, ok := in.System.(string); ok && s != "" {
		messages = append(messages, map[string]any{"role": "system", "content": s})
	}
	for _, m := range in.Messages {
		msg, ok := m.(map[string]any)
		if !ok {
			continue
		}
		messages = append(messages, map[string]any{
			"role":    msg["role"],
			"content": toOpenAIContent(msg["content"]),
		})
	}

	body := map[string]any{
		"model":      targetModel,
		"messages":   messages,
		"max_tokens": in.MaxTokens,
		"stream":     in.Stream,
	}
	if in.Tools != nil {
		body["tools"] = in.Tools
	}
	for _, k := range []string{"temperature", "top_p", "frequency_penalty", "presence_penalty"} {
		if v, ok := in.Extra[k]; ok {
			body[k] = v
		}
	}
	return body
}

// ── 内容块归一化 ──────────────────────────────

func normalizeOpenAIContent(content any) []any {
	if s, ok := content.(string); ok {
		return []any{map[string]any{"type": "text", "text": s}}
	}
	arr, ok := content.([]any)
	if !ok {
		return []any{}
	}
	out := make([]any, 0, len(arr))
	for _, b := range arr {
		block, ok := b.(map[string]any)
		if !ok {
			out = append(out, b)
			continue
		}
		switch block["type"] {
		case "text":
			// 兼容客户端发 text 为 JSON 数字（float64）的情况，统一转为字符串
			out = append(out, map[string]any{"type": "text", "text": fmt.Sprintf("%v", block["text"])})
		case "image_url":
			out = append(out, openAIImageToAnthropic(block["image_url"]))
		default:
			out = append(out, block)
		}
	}
	return out
}

func normalizeResponsesContent(content any) []any {
	if s, ok := content.(string); ok {
		return []any{map[string]any{"type": "text", "text": s}}
	}
	arr, ok := content.([]any)
	if !ok {
		return []any{}
	}
	out := make([]any, 0, len(arr))
	for _, b := range arr {
		block, ok := b.(map[string]any)
		if !ok {
			out = append(out, b)
			continue
		}
		switch block["type"] {
		case "input_text", "output_text":
			out = append(out, map[string]any{"type": "text", "text": fmt.Sprintf("%v", block["text"])})
		case "input_image":
			out = append(out, responsesImageToAnthropic(block))
		default:
			out = append(out, block)
		}
	}
	return out
}

func openAIImageToAnthropic(imageURL any) map[string]any {
	var url string
	switch v := imageURL.(type) {
	case string:
		url = v
	case map[string]any:
		url = getString(v, "url")
	}
	// data:<media_type>;base64,<data>
	if strings.HasPrefix(url, "data:") {
		if i := strings.Index(url, ";base64,"); i > 5 {
			mediaType := url[5:i]
			data := url[i+len(";base64,"):]
			return map[string]any{"type": "image", "source": map[string]any{"type": "base64", "media_type": mediaType, "data": data}}
		}
	}
	return map[string]any{"type": "image", "source": map[string]any{"type": "url", "url": url}}
}

func responsesImageToAnthropic(block map[string]any) map[string]any {
	if iu, ok := block["image_url"]; ok && iu != nil {
		return openAIImageToAnthropic(iu)
	}
	if u, ok := block["url"]; ok && u != nil {
		return map[string]any{"type": "image", "source": map[string]any{"type": "url", "url": u}}
	}
	return block
}

func toOpenAIContent(content any) any {
	arr, ok := content.([]any)
	if !ok {
		return content
	}
	// 全是文本块则合并为字符串（对齐 Node）
	allText := true
	for _, b := range arr {
		if block, ok := b.(map[string]any); !ok || block["type"] != "text" {
			allText = false
			break
		}
	}
	if allText {
		parts := make([]string, 0, len(arr))
		for _, b := range arr {
			parts = append(parts, getString(b.(map[string]any), "text"))
		}
		return strings.Join(parts, "\n")
	}
	out := make([]any, 0, len(arr))
	for _, b := range arr {
		block, ok := b.(map[string]any)
		if !ok {
			out = append(out, b)
			continue
		}
		switch block["type"] {
		case "text":
			out = append(out, map[string]any{"type": "text", "text": block["text"]})
		case "image":
			src, _ := block["source"].(map[string]any)
			var url string
			if getString(src, "type") == "base64" {
				url = "data:" + getString(src, "media_type") + ";base64," + getString(src, "data")
			} else {
				url = getString(src, "url")
			}
			out = append(out, map[string]any{"type": "image_url", "image_url": map[string]any{"url": url}})
		default:
			out = append(out, block)
		}
	}
	return out
}

// extractText 提取 Anthropic content 数组里的纯文本拼接。
func extractText(content any) string {
	arr, ok := content.([]any)
	if !ok {
		if content == nil {
			return ""
		}
		return getStringValue(content)
	}
	var sb strings.Builder
	for _, b := range arr {
		if block, ok := b.(map[string]any); ok && block["type"] == "text" {
			sb.WriteString(getString(block, "text"))
		}
	}
	return sb.String()
}

// systemTextFromContent 提取 OpenAI system 消息文本（string 或数组首块）。
func systemTextFromContent(content any) string {
	if s, ok := content.(string); ok {
		return s
	}
	if arr, ok := content.([]any); ok && len(arr) > 0 {
		if block, ok := arr[0].(map[string]any); ok {
			return getString(block, "text")
		}
	}
	return ""
}

// shortID 生成 8 位 hex，对应 Node 的 randomUUID().slice(0,8)。
func shortID() string {
	b := make([]byte, 4)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

// ── map 取值辅助 ──────────────────────────────

func getString(m map[string]any, key string) string {
	if m == nil {
		return ""
	}
	s, _ := m[key].(string)
	return s
}

func getStringValue(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func getBool(m map[string]any, key string) bool {
	b, _ := m[key].(bool)
	return b
}

// getIntDefault JSON 数字默认解析为 float64，做兼容转换。
func getIntDefault(m map[string]any, key string, def int) int {
	switch v := m[key].(type) {
	case float64:
		return int(v)
	case int:
		return v
	}
	return def
}

func getSlice(m map[string]any, key string) []any {
	s, _ := m[key].([]any)
	return s
}
